sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,293 @@
+        {
+            int streamLength = 0;
+            if (stream.CanSeek)
+            {
+                long longLength = stream.Length - stream.Position;
+                if (longLength > int.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException(
+                        nameof(stream),
+                        "Stream length must be less than Int32.MaxValue");
+                }
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
+            if (async)
+            {
+                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.CopyTo(memoryStream);
+            }
+            return new BinaryData(data: memoryStream.GetBuffer().AsMemory(0, (int)memoryStream.Position), ownsBuffer: true);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object using
+        /// the <see cref="JsonSerializer"/>.
+        /// </summary>
+        ///
+        /// <typeparam name="T">The type to use when serializing the data.</typeparam>
+        /// <param name="jsonSerializable">The data to use.</param>
+        /// <param name="options">The options to use when serializing to JSON.</param>
+        ///
+        /// <returns>A value representing the UTF-8 encoding of the JSON representation of <paramref name="jsonSerializable" />.</returns>
+        public static BinaryData FromObjectAsJson<T>(
+            T jsonSerializable,
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+            JsonSerializerOptions? options = default)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+        {
+            byte[] buffer = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, typeof(T), options);
+            return new BinaryData(buffer);
+        }
+
+        /// <summary>
+        /// Converts the value of this instance to a string using UTF-8.
+        /// </summary>
+        /// <returns>
+        /// A string from the value of this instance, using UTF-8 to decode the bytes.
+        /// </returns>
+        public override string ToString()
+        {
+            if (MemoryMarshal.TryGetArray(
+                _bytes,
+                out ArraySegment<byte> data))
+            {
+                return Encoding.UTF8.GetString(data.Array, data.Offset, data.Count);
+            }
+            return Encoding.UTF8.GetString(_bytes.ToArray());
+        }
+
+        /// <summary>
+        /// Converts the <see cref="BinaryData"/> to a read-only stream.
+        /// </summary>
+        /// <returns>A stream representing the data.</returns>
+        public Stream ToStream() =>
+            new ReadOnlyMemoryStream(_ownsBuffer ? _bytes : _bytes.ToArray());
```

Comment:
Observation: _\_ownsBuffer_ being true implies that _\_bytes_ is a `byte[]`, and in all other cases `ToArray` results in a `byte[]` anyway. In theory you could drop `ReadOnlyMemoryStream` entirely and use the built-in `MemoryStream` class, passing not-writable / not-exposable to the ctor. However, doing so runs the risk that callers might take a dependency on this implementation detail and simply cast the return value to `MemoryStream` once this method returns.

I don't know if I'd recommend changing this code, just wanted to point this out so that you could delete a bunch of unneeded internal implementation if you were ok with the trade-offs involved.

Reply:
> In theory you could drop ReadOnlyMemoryStream entirely and use the built-in MemoryStream class, passing not-writable / not-exposable to the ctor

Good point.. I think part of the benefit to using ReadOnlyMemoryStream is that users won't be able to depend on us using MemoryStream under the hood, so even though we don't have the perf need it still might be worth leaving in?

Reply:
@bartonjs @KrzysztofCwalina Do you have thoughts on this? Do you think there's risk of callers taking a dependency on the concrete type that's returned, even if the return parameter is typed as the base class `Stream`?

Reply:
Actually, wouldn't it still be ROM here - https://github.com/JoshLove-msft/azure-sdk-for-net/blob/bd-system/sdk/core/System.BinaryData/src/BinaryData.cs#L183

Reply:
No, `MemoryStream.GetBuffer()` returns a `byte[]`, so you'd still end up with a `ROM<byte>` backed by a \[potentially-sliced\] `byte[]`.

Reply:
>  Do you think there's risk of callers taking a dependency on the concrete type that's returned, even if the return parameter is typed as the base class Stream?

Absolutely.  Happens all the time. That's why we advocate just being honest with your return type.

Or, as  Anthony Moore (as quoted in Framework Design Guidelines, 3rd edition) said:

> In design meetings, we would often describe this set of guidelines with the statement that you should “require the weakest thing you need, and return the strongest thing you have.”

Reply:
I guess a better question would be: Say that we returned an instance of the public `MemoryStream` class instead of our internal `ReadOnlyMemoryStream` class, but the public return type was still typed as the base class `Stream`. If in a future release we change our concrete return type from `MemoryStream` to `<SomeOtherStream>`, it would obviously have the potential to break people. But would we care, or is this the kind of change that would be forbidden in a future version?

Reply:
Going from a non-public type to a public type isn't breaking (unless the behaviors were wildly different); so I'd just leave this using the private stream type.

Reply:
Ok will leave it as is.

Reply:
> But would we care, or is this the kind of change that would be forbidden in a future version?

In .NET Framework we'd've forbidden it.  In .NET (Core) and/or an OOB we haven't yet found where the boundary is... probably "extremely hesitant".

It's in FDGv3 (D.12.3 Changing the Type of Values Returned from a Method)

> The .NET BCL team is generally supportive of returning a “more
derived” type than in a previous version, unless there’s a strong reason to
believe type equality checks will be routinely used for that particular
method. Returning a value that would break existing “hard-casting”
callers is something the .NET BCL team generally considers an unacceptable
breaking change.